unlock-app/src/middlewares/providerMiddleware.ts

```diff
@@ -111,26 +74,28 @@ export async function initializeUnlockProvider(
   }
 }
 
-export const providerMiddleware = (config: any) => {
-  return ({ getState, dispatch }: { [key: string]: any }) => {
-    return function(next: any) {
-      // Initialize provider based on the one grabbed in the state. Fragile?
-      setTimeout(() => {
-        const provider = config.providers[getState().provider]
-        initializeProvider(provider, dispatch)
-      }, 0)
-
-      return function(action: Action) {
-        const providerName = getState().provider
-        const provider = config.providers[providerName]
-        if (action.type === SET_PROVIDER) {
-          // Only initialize the provider if we haven't already done so.
-          if (action.provider !== providerName) {
-            const newProvider = config.providers[action.provider]
-            initializeProvider(newProvider, dispatch)
-          }
-        } else if (action.type === GOT_ENCRYPTED_PRIVATE_KEY_PAYLOAD) {
-          initializeUnlockProvider(action, provider, dispatch)
+/**
+ *
+ * @param config
+ * @param getProvider
+ */
+export const providerMiddleware = (
+  config: any,
+  getProvider: () => UnlockProvider,
+  setProvider: (provider: UnlockProvider) => void
+) => {
+  const { readOnlyProvider, requiredNetworkId } = config
+  return ({ dispatch }: { [key: string]: any }) => {
+    return function middleware(next: any) {
+      return function handler(action: Action) {
+        const provider = getProvider()
+        if (action.type === GOT_ENCRYPTED_PRIVATE_KEY_PAYLOAD) {
+          const unlockProvider = new UnlockProvider({
```

Comment:
When we get an encrypted key, we instantiate a new unlock provider. Eventually we could move that logic to the login page.